### ISSUE TITLE
Change Request: allow `allowReserved` regardless of version

### ISSUE THREAD

**Issue by aladdin-add** [eslint/eslint#19278]
### Discussed in https://github.com/eslint/eslint/discussions/19256

<div type='discussions-op-text'>

<sup>Originally posted by **jsumners-nr** December 19, 2024</sup>
### ESLint version

9.17.0

### What problem do you want to solve?

Per this difficult to find reference https://eslint.org/docs/latest/use/configure/language-options#specifying-parser-options, the `allowReserved` option is tied to the ECMA v3 setting. I very much have a case where I need to test that reserved words do not interfere with operations. As a result, I seem to be completely blocked on being able to upgrade to eslint v9.

https://github.com/newrelic/node-newrelic/blob/47b8398820d665a85a96ae84e30eaaf20564dcf8/test/unit/instrumentation/connect.test.js#L130-L140

### What do you think is the correct solution?

Allow us to turn off this not very helpful error.

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

_No response_</div>

**Comment by mdjermanovic** [comment]
Can you please clarify the request - what is the use case for allowing the use of reserved words as identifiers with ecmaVersion > 3?

> As a result, I seem to be completely blocked on being able to upgrade to eslint v9.

What exactly has changed in ESLint v9 in this regard?

**Comment by jsumners-nr** [comment]
😒 ... follow a completely new issue.


> Can you please clarify the request - what is the use case for allowing the use of reserved words as identifiers with ecmaVersion > 3?

That is very clearly outlined in the original issue (https://github.com/eslint/eslint/issues/19251).

> What exactly has changed in ESLint v9 in this regard?

I do not know. All I know is that the exact linked case was not failing prior to v9. As a workaround, I wrapped the offending code in a `new Function` -- https://github.com/jsumners-nr/node-newrelic/blob/af3771daf9f64768d628ab8cf27599c5defa0354/test/unit/instrumentation/connect.test.js#L126-L135

**Comment by mdjermanovic** [comment]
> > Can you please clarify the request - what is the use case for allowing the use of reserved words as identifiers with ecmaVersion > 3?
> 
> That is very clearly outlined in the original issue ([#19251](https://github.com/eslint/eslint/issues/19251)).

I see only a link to a test case, but I'm not sure what it is testing. Can you please clarify why `static` should be allowed as an identifier? In the runtime, it produces a syntax error:

```js
"use strict";

function static () {};

```

```
function static () {};
         ^^^^^^

SyntaxError: Unexpected strict mode reserved word
```

**Comment by mdjermanovic** [comment]
Also note that in the new config system .js files are parsed as ESM modules by default. By your test file, I suspect that it wasn't intended to be an ESM module, but a script or CommonJS module, so updating `languageOptions.sourceType` to `"script"` or `"commonjs"` in your eslint configuration might fix the problem because `static` isn't a reserved word in non-strict mode code (ESM modules are always strict mode code).

https://eslint.org/docs/latest/use/configure/language-options#specifying-javascript-options

**Comment by jsumners-nr** [comment]
> Can you please clarify why `static` should be allowed as an identifier?

As far as I know, it's only to guarantee that our module works with code that does such silly things.

> .js files are parsed as ESM modules by default.

A poor assumption. Particularly since code bases can be mixed, as the one that generated this issue is.

> so updating `languageOptions.sourceType` to `"script"` or `"commonjs"` in your eslint configuration might fix the problem because `static` isn't a reserved word in non-strict mode code

Presumably this is why it wasn't a problem in `eslint@8`.

**Comment by mdjermanovic** [comment]
> > Can you please clarify why `static` should be allowed as an identifier?
> 
> As far as I know, it's only to guarantee that our module works with code that does such silly things.

Code with identifiers named `static` wouldn't work as an ESM module in runtime. The solution for linting such code would be to set `languageOptions.sourceType` to `"script"` or `"commonjs"` since these are the only contexts in which such code can be syntactically valid.

So I think no changes should be made regarding the `allowReserved` option, i.e., it should only apply when `ecmaVersion` is `3` as decided in the [2021-11-18 TSC meeting](https://github.com/eslint/tsc-meetings/blob/main/notes/2021/2021-11-18.md).

@eslint/eslint-team any other opinions?

**Comment by aladdin-add** [comment]
I agreed.
